File: charging_session_service/src/models/charging_session.model.js
import mongoose from 'mongoose';
import { ChargingSessionStatus } from '../constants/session.constants';

const { Schema } = mongoose;

const ChargingSessionSchema = new Schema({

    id: {
        type: String,
        required: true,
        unique: true,
    },
    booking_id: {
        type: Schema.Types.ObjectId,
        ref: 'Booking',
        required: true,
    },
     session_code: {
        type: String,
        required: true,
        unique: true,
    },
    vehicle_name: {
        type: String,
        required: true,
    },
    vehicle_number: {
        type: String,
        required: true,
    },
    start_time: {
        type: Date,
        required: true,
    },
    end_time: {
        type: Date,
        required: true,
    },
    duration_time: {
        type: Number,
        required: true,
        min: 0,
    },
    start_soc_percent: {
        type: Number,
        min: 0,
        max: 100,
        required: true,
    },
    end_soc_percent: {
        type: Number,
        min: 0,
        max: 100,
        required: true,
    },
    status: {
        type: String,
        enum: Object.values(ChargingSessionStatus),
        default: ChargingSessionStatus.IN_PROGRESS,
        required: true,
    },
    total_kwh: {
        type: Number,
        required: true,
        min: 0,
    },
    total_price: {
        type: Number,
        required: true,
        min: 0,
    },
    payment_method: {
        type: String,
        required: true,
    },
    staff_operation: {
        type:String,
        required: false,
    },
    created_at: {
        type: Date,
        default: Date.now,
    },
    updated_at: {
        type: Date,
        default: Date.now,
    },
}, {
    timestamps: true,
    collection: 'charging_sessions'
});

export default mongoose.model('ChargingSession', ChargingSessionSchema);

File: ev_point_session/lib/features/charging_session/data/models/charging_session_model.dart
import 'package:ev_point_session/features/charging_session/domain/entities/charging_session.dart';

class ChargingSessionModel extends ChargingSession {
  ChargingSessionModel({
    required super.id,
    required super.bookingId,
    required super.sessionCode,
    required super.vehicleName,
    required super.vehicleNumber,
    required super.startTime,
    required super.endTime,
    super.durationTime,
    required super.status,
    required super.startSoCPercent,
    required super.endSoCPercent,
    required super.totalKwh,
    required super.totalAmount,
    required super.paymentMethod,
    required super.staffOperation,
    required super.createdAt,
    super.updatedAt,
  });

  factory ChargingSessionModel.fromJson(Map<String, dynamic> json) {

    final String id = json['_id']?.toString() ?? '';

    const int bookingId = 0; 

    DateTime parseDate(dynamic value, {DateTime? fallback}) {
      if (value == null) return fallback ?? DateTime.now();
      if (value is String && value.isNotEmpty) {
        return DateTime.tryParse(value) ?? (fallback ?? DateTime.now());
      }
      return fallback ?? DateTime.now();
    }

    DateTime? parseNullableDate(dynamic value) {
      if (value == null) return null;
      if (value is String && value.isNotEmpty) {
        return DateTime.tryParse(value);
      }
      return null;
    }

    return ChargingSessionModel(
      id: id,
      bookingId: bookingId,
      sessionCode: json['session_code']?.toString() ?? '',
      vehicleName: json['vehicle_name']?.toString() ?? '',
      vehicleNumber: json['vehicle_number']?.toString() ?? '',
      startTime: parseDate(json['start_time']),
      endTime: parseDate(json['end_time'], fallback: DateTime.now()),
      durationTime: null,
      status: json['status']?.toString() ?? '',
      startSoCPercent: json['start_soc_percent']?.toString() ?? '',
      endSoCPercent: json['end_soc_percent']?.toString() ?? '',
      totalKwh: (json['total_kwh'] is num)
          ? (json['total_kwh'] as num).toDouble()
          : double.tryParse(json['total_kwh']?.toString() ?? '0') ?? 0,
      totalAmount: (json['total_price'] is num)
          ? (json['total_price'] as num).toDouble()
          : double.tryParse(json['total_price']?.toString() ?? '0') ?? 0,
      paymentMethod: json['payment_method']?.toString() ?? '',
      staffOperation: json['staff_operation']?.toString() ?? '',
      createdAt: parseDate(json['createdAt']),
      updatedAt: parseNullableDate(json['updatedAt']),
    );
  }
}
